### PR TITLE
Muse-Glimmer - text-config torchvision fix + HuggingFace StateDictAdapter

### PR DESCRIPTION
**Pull request by AlirezaShamsoshoara**
## Summary

This PR makes the newly-landed **Muse-Glimmer** model usable for public, text-only
users and adds HuggingFace checkpoint interop:

1. **Fix:** `muse_glimmer/config_registry.py` imported `torchvision` (an *optional*
   dependency) at module top level, which broke the **text-only** configs
   (`muse_glimmer_debugmodel`, `muse_glimmer_30b`) for anyone without torchvision
   installed. Made the multimodal imports lazy.
2. **Feature:** added `MuseGlimmerStateDictAdapter`, mapping torchtitan's native
   FQNs ↔ the released HuggingFace `meta-models/Muse-Glimmer-30B` layout, and wired
   it into the model spec (`state_dict_adapter=` was `None`). This enables
   `checkpoint.last_save_in_hf` / `initial_load_in_hf` and HF-seeded TitanRL.

The weight mapping is **verified against the released checkpoint**: DCP-loaded weights
are bitwise-equal to the permuted HF weights, and **end-to-end logit parity passes**
(0.99996 cosine, 96.9% argmax) - see Sanity checks. 

---

## Changes

| File | Type | Lines | What |
|---|---|---|---|
| `torchtitan/models/muse_glimmer/config_registry.py` | modified | +15/−6 | Move `MMDataLoader` / `resize_to_pixel_budget` imports inside `_muse_glimmer_mm_dataloader()` so text configs load without torchvision (optional dep). |
| `torchtitan/models/muse_glimmer/state_dict_adapter.py` | **new** | 201 | `MuseGlimmerStateDictAdapter` (native ↔ HF Gemma3-style FQNs; Q/K RoPE permute; skips vision keys). |
| `torchtitan/models/muse_glimmer/__init__.py` | modified | +2/−1 | Import the adapter; set `state_dict_adapter=MuseGlimmerStateDictAdapter` in `model_registry`. |

No other files touched.

---

## Environment / package versions used to develop & validate

Reproducible **uv** venv (no conda required):

| package | version |
|---|---|
| python | 3.12 |
| torch | 2.14.0.dev20260810+cu130 |
| torchvision | 0.29.0.dev20260811+cu130 (only needed for `*_mm` / parity env) |
| triton | 3.8.0+git675c5987 |
| transformers | 5.15.0 (ships `muse_glimmer` natively) |
| safetensors | 0.8.0 |
| huggingface_hub | 1.27.0 |
| accelerate | 1.14.0 |
| wandb | 0.28.1 |
| torchdata | 0.11.0 |
| datasets | 4.7.0 |
| tokenizers | 0.22.2 |
| spmd_types | 0.2.1 |
| numpy | 2.5.2 |

Hardware: 8× NVIDIA H100 (95 GiB), driver 580.126.09, CUDA 13.0, cuDNN 9.24.

---

### Build the venv
```bash
uv venv --python 3.12 .venv-mg && source .venv-mg/bin/activate
uv pip install --prerelease=allow --index-url https://download.pytorch.org/whl/nightly/cu130 \
  "torch==2.14.0.dev20260810+cu130" "torchvision==0.29.0.dev20260811+cu130"
uv pip install "torchdata>=0.8.0" "datasets>=3.6.0,<4.8.0" tensorboard wandb \
  "tyro>=1.0.5" "tokenizers>=0.15.0" safetensors einops pillow "spmd_types==0.2.1"
# for HF parity only:
uv pip install transformers accelerate
```

### Train smoke tests
```bash
export PYTHONPATH="$PWD"
# 1 GPU
NGPU=1 MODULE=muse_glimmer CONFIG=muse_glimmer_debugmodel ./run_train.sh \
  --metrics.enable_wandb --dataloader.dataset=c4_test --hf_assets_path=./tests/assets/tokenizer
# 8 GPU, 30B, FSDP8
NGPU=8 MODULE=muse_glimmer CONFIG=muse_glimmer_30b ./run_train.sh \
  --metrics.enable_wandb --dataloader.dataset=c4_test --hf_assets_path=./tests/assets/tokenizer \
  --parallelism.data_parallel_shard_degree=8 --training.local_batch_size=1 --training.steps=8
```
image
image

### Download the released HF checkpoint (for parity)
```bash
# 59.6 GB, 2 shards. (hf_hub Xet transport hit an httpx/IPv6-proxy bug on this host;
#  plain curl works fine.)
BASE="https://huggingface.co/meta-models/Muse-Glimmer-30B/resolve/main"
for f in config.json model.safetensors.index.json tokenizer.json tokenizer_config.json \
         model-00001-of-00002.safetensors model-00002-of-00002.safetensors; do
  curl -sSL --fail --retry 5 -C - "$BASE/$f" -o "hf_ckpt/$f"
done
```

---

## Sanity checks & tests (all commands run; all green)

(test .py and .sh files are not included in this PR, but can be shared internally)

Cap threads to avoid CPU oversubscription: `export OMP_NUM_THREADS=4`.

| # | Test (script) | What it asserts | Result |
|---|---|---|---|
| 1 | `test_sd_adapter_roundtrip.py` | `from_hf(to_hf(sd))` bitwise on all 99 debug params; every key mapped; Q/K permute applied & exact inverse | **PASS** |
| 2 | `test_sd_adapter_hf_parity.py` | adapter's HF key set **== the 627 text-decoder keys** in the released `model.safetensors.index.json` (0 uncovered, 0 invented) | **PASS** |
| 3 | `test_sd_adapter_safetensors.py` | write real `.safetensors`+index via `to_hf`, read back via `from_hf` → bitwise-equal, 0 missing/unexpected | **PASS** |
| 4 | `test_sd_adapter_hf_e2e.sh` | 2-GPU FSDP through the real `CheckpointManager`: `last_save_in_hf` writes HF safetensors; `initial_load_in_hf` reloads & continues loss trajectory | **PASS** (p1=0,p2=0) |
| 5 | `test_loaded_weight_check.py` | loaded `wq` **== `reverse_permute(HF q_proj)` bitwise (max_abs_diff 0.0)**; vs raw-HF cosine 0.018 | **PASS** |
| 6 | `test_parity_{hf,titan}_stage.py` (real 59.6 GB ckpt) | full HF-vs-torchtitan forward, 256 tokens: **global cosine 0.99996, per-token min 0.99926, argmax 96.9%, top-5 96.5%** | **PASS** |
| 7 | regression (`logs/46_v2_regression.log`) | training unchanged with adapter wired in | **PASS** |
| 8 | text-config import without torchvision (uv `.venv-mg-notv`) | `import config_registry` + build debug/30B succeed; MM path still correctly requires torchvision | **PASS** |

**Parity evidence (adapter + torchtitan forward both correct):**
- Final end-to-end logits: **global cosine 0.99996**, per-token cosine mean 0.99994 / min 0.99926,
  argmax agreement 96.9%, top-5 overlap 96.5% (remaining gap is bf16 forward noise).
- token-embedding cosine **1.00000**; DCP-loaded `wq` **== `reverse_permute(HF q_proj)` bitwise (diff 0.0)**.
- Isolated attention scores (q_proj→qk_norm→scale→RoPE→scores), full GQA (32Q/2KV), multi-position,
  with the permute: cosine **1.0** (max_abs_diff 7.6e-5).

---

### CODE REVIEW

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/muse_glimmer/state_dict_adapter.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        # All Muse Glimmer layers carry a ComplexRoPE config (NoPE layers still
+        # build one; RoPE is guarded in forward), so this validates cleanly.
+        self._validate_hf_rope_config(ComplexRoPE.Config)
+
+        n_heads, n_kv_heads, dim, head_dim = self._attn_geometry()
+        state_dict: dict[str, Any] = {}
+        for key, value in hf_state_dict.items():
+            # Skip multimodal (vision tower/adapter/projection) weights -- this
+            # adapter handles the text decoder; the vision path is loaded
+            # separately by the MM model spec when enabled.
```
Comment (fegin) [inline torchtitan/models/muse_glimmer/state_dict_adapter.py]: Is this True?

**Review by fegin — CHANGES_REQUESTED** [review]
Is this text only? If it is, we should gate the model type upfront to assert that if the model is mm.

File: torchtitan/models/muse_glimmer/state_dict_adapter.py (hunk 1 above)
Comment (AlirezaShamsoshoara) [reply torchtitan/models/muse_glimmer/state_dict_adapter.py]: @fegin Thanks for catching this. This is not true! I was trying to only check the forward parity only in text decoder part (rather than vision) and I intentionally did not include the vision. I added a commit now to support vision. Please recheck the `state_dict_adapter` file.

**Comment by AlirezaShamsoshoara** [comment]
@fegin  Adding the mm / vision support to state_dict_adapter:

The adapter was text-only for a VLM. First response was to gate multimodal configs with `NotImplementedError`. Follow-up decision: **align with the repo's VLM convention** - `qwen3_5` and `kimi_k2_7` (the other torchtitan models that own a vision encoder) **map vision weights in their state_dict_adapter**. So `MuseGlimmerStateDictAdapter` was extended to map the full checkpoint (text decoder + vision tower/adapter/projection), and the gate was removed.


## Vision mapping (native <-> HF)

HF prefixes: `model.vision_tower.` (encoder), `model.vision_adapter.`, `model.vision_projection`.

| torchtitan native | HF | transform |
|---|---|---|
| `vision_encoder.conv1_linear.weight` | `vision_tower.patch_embedder.patch_embedding.weight` | none (Linear both sides, [1536,1176]) |
| `vision_encoder.positional_embedding_vlm` | `vision_tower.patch_embedder.position_embedding_table.weight` | none ([1024,1536]) |
| `vision_encoder.ln_pre/ln_post.{weight,bias}` | `vision_tower.ln_pre/ln_post.*` | none |
| `vision_encoder.layers.{}.attn.wq/wk.{weight,bias}` | `vision_tower.layers.{}.attn.q_proj/k_proj.*` | **RoPE permute** (q/k, weight+bias) |
| `vision_encoder.layers.{}.attn.wv/proj.{weight,bias}` | `vision_tower.layers.{}.attn.v_proj/proj.*` | none |
| `vision_encoder.layers.{}.mlp.linear_fc1/fc2.{weight,bias}` | `vision_tower.layers.{}.mlp.fc1/fc2.*` | none |
| `vision_encoder.layers.{}.norm1/norm2.{weight,bias}` | `vision_tower.layers.{}.norm1/norm2.*` | none |
| `vision_adapter.c_fc.weight` | `vision_adapter.fc1.weight` | none |
| `vision_adapter.c_proj.weight` | `vision_adapter.fc2.weight` | none |
| `vision_projection.weight` | `vision_projection.weight` | none |

## Sanity checks & tests

Cap threads to avoid CPU oversubscription: `export OMP_NUM_THREADS=4`.

Continue from the previous table:

| # | Test (script) | What it asserts | Result |
|---|---|---|---|
| 9 | full key coverage (`test_full_coverage.py`, 30B_mm) | adapter maps **all 1436** HF keys (627 text + 809 vision), **0 uncovered / 0 invented** | **PASS** |
| 10 | MM round-trip (`test_mm_roundtrip.py`) | `from_hf(to_hf(sd))` bitwise on all 156 debug_mm params (99 text + 57 vision); vision q/k permute (weight+bias) exact inverse | **PASS** |
| 11 | real MM weights load (30B_mm, DCP) | all 1436 real HF tensors load, 0 missing/unexpected; vision `wq == reverse_permute(HF q_proj)` bitwise (diff 0.0) | **PASS** |
| 12 | vision RoPE permute (isolated q·k, real 30B vision geom) | with permute, vision attention scores match HF **cosine 0.9999998**; without, 0.89 | **PASS** |


## Vision validation (all PASS)

| Test | What it proves | Result |
|---|---|---|
| Full key coverage (`test_full_coverage.py`, 30B_mm) | adapter maps **all 1436** HF keys (627 text + 809 vision), **0 uncovered / 0 invented** | PASS |
| MM round-trip (`test_mm_roundtrip.py`) | `from_hf(to_hf(sd))` bitwise on all **156** debug_mm params (99 text + 57 vision); q/k permute (weight+bias) is an exact inverse | PASS |
| Real-weights load (30B_mm, DCP) | all 1436 real HF tensors load, **0 missing / 0 unexpected**; loaded vision `wq == reverse_permute(HF q_proj)` **bitwise (max_abs_diff 0.0)** | PASS |
| Vision RoPE permute (isolated q.k, real 30B geom: 16 heads, hd=96, 2D rope) | with permute, vision attention scores match HF **cosine 0.9999998** (max_abs_diff 1.9e-6); without permute, 0.89 | PASS |
| Text parity unaffected | text round-trip still 99/99 bitwise | PASS |

File: torchtitan/models/muse_glimmer/state_dict_adapter.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        if dim1 is None:
+            dim1 = w.shape[0]
+        if dim2 is None:
+            dim2 = w.shape[1]
+        return (
+            w.view(n_heads_arg, dim1 // n_heads_arg // 2, 2, dim2)
+            .transpose(1, 2)
+            .reshape(dim1, dim2)
+            .clone()
+        )
```
Comment (fegin) [inline torchtitan/models/muse_glimmer/state_dict_adapter.py:189]: CLAUDE verified it locally
```
                                                dp_shard=2   =4     =8
  debugmodel wq  [256, 256]    n_heads=4          OK         OK    FAIL
  debugmodel wk  [128, 256]    n_kv_heads=2       OK        FAIL   FAIL
  30B wq         [4096, 6656]  n_heads=32         OK         OK     OK
  30B wk         [256, 6656]   n_kv_heads=2       OK        FAIL   FAIL
```

Suggested fix:
```
  def _permute(self, w, n_heads_arg, dim1=None, dim2=None):
      ...
      if isinstance(w, DTensor):
          orig = w.placements
          w = w.redistribute(placements=[Replicate()] * w.device_mesh.ndim)
          out = _permute_local(w, n_heads_arg, dim1, dim2)
          return out.redistribute(placements=orig)
      return _permute_local(w, n_heads_arg, dim1, dim2)
```

**Review by fegin — COMMENTED** [review]
Two issues found by CLAUDE:

  - Issue #1 (blocker): to_hf crashes at dp_shard > n_kv_heads (=2). Confirmed end-to-end on a fully_shard'd debugmodel. Affects save and load. Now has a verified fix.
  - Issue #2 (blocker): 57 of 156 tensors — the whole vision stack — silently dropped for *_mm flavors. Kimi raises on this exact case; qwen3_5 maps it.

File: torchtitan/models/muse_glimmer/state_dict_adapter.py (hunk 2 above)
Comment (AlirezaShamsoshoara) [reply torchtitan/models/muse_glimmer/state_dict_adapter.py:189]: Thanks @fegin Let me check and get back to you on this

Comment (AlirezaShamsoshoara) [reply torchtitan/models/muse_glimmer/state_dict_adapter.py:189]: Thanks @fegin for the review. 

I think the earlier commit I had [549f813](https://github.com/pytorch/torchtitan/pull/4118/commits/549f813439d6e957e64fa598b2ead464e79ebc49) should have fixed the issue 2. Please correct me if I am wrong. 

Regarding issue 1, it was valid, I tried to fix it in [e01edae](https://github.com/pytorch/torchtitan/pull/4118/commits/e01edae2487ce8ed92c46619cae7072f3293f72c)

in the live save/load path, `to_hf`/`from_hf` receive DTensors that FSDP shards along dim 0 (the q/k output rows), and the head-splitting `view(n_heads, ...)` cannot unflatten an unevenly-sharded dim, raising `Cannot unflatten unevenly sharded tensor`. My earlier round-trip/parity tests used plain (unsharded) tensors, so they never exercised the FSDP DTensor path.

Fix: wrapped the four q/k permute helpers with a `_dtensor_safe` decorator: if the input is a DTensor, redistribute to `Replicate`, run the permute on the full **local** tensor (`.to_local()`), then wrap back and redistribute to the original placements (plain tensors take the fast path). This uses the redistribute-to-Replicate strategy you suggested, packaged as one decorator across all four helpers (there are four: q/k weight + vision q/k bias), with the actual reshape done on the local tensor so the `view`/unflatten never runs through the DTensor dispatcher. Confirmed bitwise-equal to the plain permute.

Reproduced the exact failure matrix, then confirmed the fix:

| tensor | shape | heads | dp_shard=2 | =4 | =8 |
|---|---|---|---|---|---|
| debugmodel wq | [256,256] | n_heads=4 | OK | OK | fixed (was FAIL) |
| debugmodel wk | [128,256] | n_kv_heads=2 | OK | fixed (was FAIL) | fixed (was FAIL) |
| 30B wk | [256,6656] | n_kv_heads=2 | OK | fixed (was FAIL) | fixed (was FAIL) |

Could you please recheck?

**Review by fegin — APPROVED** [review]
